cpp/ql/test/library-tests/ir/ir/raw_consistency.expected

```diff
@@ -1,4 +1,64 @@
 missingOperand
+| ir.cpp:809:7:809:13 | IndirectMayWriteSideEffect: reuse of temporary object | Instruction 'IndirectMayWriteSideEffect' is missing an expected operand with tag 'Address' in function '$@'. | ir.cpp:799:6:799:25 | void HierarchyConversions() | void HierarchyConversions() |
```

Comment:
Not sure what's going on here. Still need to investigate.

Reply:
Some IR is generated for the destructor calls, which is what's causing this. I did not expect it to do that.

Reply:
With this change the problem seems to shift, and I get instructions without successors:
```
+| ir.cpp:809:3:809:3 | Chi: b | Instruction 'Chi: b' has no successors in function '$@'. | ir.cpp:799:6:799:25 | void HierarchyConversions() | void HierarchyConversions() |
+| ir.cpp:1372:23:1372:27 | BufferReadSideEffect: (reference to) | Instruction 'BufferReadSideEffect: (reference to)' has no successors in function '$@'. | ir.cpp:1367:6:1367:21 | void temporary_string() | void temporary_string() |
+| ir.cpp:1386:5:1386:15 | Chi: call to acceptValue | Instruction 'Chi: call to acceptValue' has no successors in function '$@'. | ir.cpp:1381:6:1381:30 | void temporary_destructor_only() | void temporary_destructor_only() |
+| ir.cpp:2021:5:2021:5 | Chi: z | Instruction 'Chi: z' has no successors in function '$@'. | ir.cpp:2019:6:2019:25 | void TernaryTestNonPodObj(bool, TernaryNonPodObj, TernaryNonPodObj, TernaryNonPodObj) | void TernaryTestNonPodObj(bool, TernaryNonPodObj, TernaryNonPodObj, TernaryNonPodObj) |
+| ir.cpp:2246:20:2246:50 | Chi: call to vector | Instruction 'Chi: call to vector' has no successors in function '$@'. | ir.cpp:2240:6:2240:19 | void ForDestructors() | void ForDestructors() |
+| smart_ptr.cpp:12:20:12:27 | Chi: temporary object | Instruction 'Chi: temporary object' has no successors in function '$@'. | smart_ptr.cpp:10:6:10:24 | void call_unique_ptr_arg(int*) | void call_unique_ptr_arg(int*) |
+| smart_ptr.cpp:19:20:19:21 | Chi: temporary object | Instruction 'Chi: temporary object' has no successors in function '$@'. | smart_ptr.cpp:17:6:17:24 | void call_shared_ptr_arg(float*) | void call_shared_ptr_arg(float*) |
+| smart_ptr.cpp:31:26:31:37 | BufferReadSideEffect: temporary object | Instruction 'BufferReadSideEffect: temporary object' has no successors in function '$@'. | smart_ptr.cpp:28:6:28:27 | void call_shared_ptr_consts() | void call_shared_ptr_consts() |
+| temp.cpp:15:14:15:35 | Chi: temporary object | Instruction 'Chi: temporary object' has no successors in function '$@'. | temp.cpp:14:6:14:14 | char temp_test() | char temp_test() |
+| temp.cpp:23:12:23:74 | Add: ... + ... | Instruction 'Add: ... + ...' has no successors in function '$@'. | temp.cpp:21:6:21:15 | char temp_test2() | char temp_test2() |
```

Reply:
@rdmarsh2 I think we need to further restrict the IR generation here, otherwise this cannot be merged without also doing all the destructor work in the IR translation.

Reply:
Looks like there's two other predicates I forgot about - I'll open a PR as soon as I verify that there's nothing else.

Reply:
This is blocking me at this point.

Reply:
Thanks for https://github.com/github/codeql/pull/15817

This mostly fixes the issue, except for destructors attached to statements that take a re-use expr as their qualifier. In can happen in the case of https://github.com/github/codeql/pull/15770#issuecomment-1971832877. A suppression for this will be specific to this branch, and it seems that I at least need to add the following in `ignoreExprAndDescendants`:
```ql
  or
  exists(Stmt parent |
    parent.getAnImplicitDestructorCall() = expr and
    expr.(DestructorCall).getQualifier() instanceof ReuseExpr
  )
}
```
However, adding the above still gives me some inconsistencies of the following form:
```
| ir.cpp:1379:1:1379:1 | NoOp: return ... | Instruction 'NoOp: return ...' has no successors in function '$@'. | ir.cpp:1367:6:1367:21 | void temporary_string() | void temporary_string() |
| temp.cpp:31:1:31:1 | NoOp: return ... | Instruction 'NoOp: return ...' has no successors in function '$@'. | temp.cpp:29:6:29:15 | void temp_test3() | void temp_test3() |
| temp.cpp:36:1:36:1 | NoOp: return ... | Instruction 'NoOp: return ...' has no successors in function '$@'. | temp.cpp:33:6:33:15 | void temp_test4() | void temp_test4() |
```
So, I apparently need a further exclusion somewhere, but it's not totally clear to me where.

Reply:
I think I found a solution in the form of https://github.com/github/codeql/pull/15770/files#diff-a8f8aa1dd299be8d67e9828691c3c44edb5e74d962b256017c8d0c856ab0bb90